controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -592,6 +596,128 @@ func (r *BareMetalHostReconciler) detachHost(prov provisioner.Provisioner, info
+	}
+
+	if image.Spec.NetworkDataName != "" {
+		secretKey := client.ObjectKey{
+			Name:      image.Spec.NetworkDataName,
+			Namespace: image.ObjectMeta.Namespace,
+		}
+		secretManager := r.secretManager(info.log)
+		networkData, err := secretManager.AcquireSecret(secretKey, info.host, false)
+		if err != nil {
+			return false, err
+		}
+		if image.Status.NetworkData.Version != networkData.GetResourceVersion() {
+			info.log.Info("network data in pre-provisioning image is out of date",
+				"latestVersion", networkData.GetResourceVersion(),
+				"currentVersion", image.Status.NetworkData.Version)
+			return false, nil
+		}
+	}
+	if image.Status.NetworkData.Name != image.Spec.NetworkDataName {
+		info.log.Info("network data location in pre-provisioning image is out of date")
+		return false, nil
+	}
+
+	for _, cond := range image.Status.Conditions {
+		if cond.Status == metav1.ConditionTrue {
+			switch metal3v1alpha1.ImageStatusConditionType(cond.Type) {
+			case metal3v1alpha1.ConditionImageReady:
+				return true, nil
+			case metal3v1alpha1.ConditionImageError:
+				info.log.Info("error building PreprovisioningImage",
+					"message", cond.Message)
+				return false, imageBuildError{cond.Message}
+			}
+		}
+	}
+
+	info.log.Info("pending PreprovisioningImage not ready")
+	return false, nil
+}
+
+func getHostArchitecture(host *metal3v1alpha1.BareMetalHost) string {
+	if host.Status.HardwareDetails != nil &&
+		host.Status.HardwareDetails.CPU.Arch != "" {
+		return host.Status.HardwareDetails.CPU.Arch
+	}
+	if hwprof, err := hardware.GetProfile(getHardwareProfileName(host)); err == nil {
+		return hwprof.CPUArch
+	}
+	return ""
+}
+
+func (r *BareMetalHostReconciler) getPreprovImage(info *reconcileInfo, formats []metal3v1alpha1.ImageFormat) (*provisioner.PreprovisioningImage, error) {
+	if formats == nil {
+		// No image build requested
+		return nil, nil
+	}
+
+	if len(formats) == 0 {
+		return nil, imageBuildError{"no acceptable formats for preprovisioning image"}
+	}
+
+	expectedSpec := metal3v1alpha1.PreprovisioningImageSpec{
+		NetworkDataName: info.host.Spec.PreprovisioningNetworkDataName,
+		Architecture:    getHostArchitecture(info.host),
+	}
+
+	preprovImage := metal3v1alpha1.PreprovisioningImage{}
+	key := client.ObjectKey{
+		Name:      info.host.Name,
+		Namespace: info.host.Namespace,
+	}
+	err := r.Get(context.TODO(), key, &preprovImage)
+	if k8serrors.IsNotFound(err) {
+		info.log.Info("creating new PreprovisioningImage")
+		preprovImage = metal3v1alpha1.PreprovisioningImage{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      key.Name,
+				Namespace: key.Namespace,
+			},
+			Spec: expectedSpec,
+		}
+		controllerutil.SetControllerReference(info.host, &preprovImage, r.Scheme())
+		err = r.Create(context.TODO(), &preprovImage)
+		return nil, err
+	}
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to retrieve pre-provisioning image data")
+	}
+
+	if !apiequality.Semantic.DeepEqual(preprovImage.Spec, expectedSpec) {
```

Comment:
The reason for this block isn't very clear to me, could you please provide some details? Shouldn't be the BMH controller take care just of the resource creation (and wait for the resource to be ready)?

Reply:
If any of the data changes (specifically, if a different network data Secret is specified), we need to update the PreprovisioningImage so that the next boot will use the latest data.

Reply:
The Preprov controller is watching the Secret for changes, but if the user changes the BMH to point to a _different_ Secret then it is the BMH controller's responsibility to update the PreprovisioningImage. The PreprovisioningImage controller never updates its own Spec.